### ISSUE TITLE
Phase 3.2: Train LoRA Model (Local or Cloud)

### ISSUE THREAD

**Issue by JayTitus** [JayTitus/Number2#10]
## Objective
Fine-tune a LoRA adapter on your writing style using your training data.

## Prerequisites
- Issue #9 complete (training data generated)
- EITHER: NVIDIA GPU with 8GB+ VRAM
- OR: Cloud GPU access (Colab, RunPod, Lambda)
- OR: OpenAI fine-tuning API access

### Option A: Local GPU (Best if you have hardware)
Requires NVIDIA GPU with 8GB+ VRAM (RTX 3070+, RTX 4060+, etc.)

### Option B: Google Colab (Free tier may work)
Use Colab's free GPU, may have limitations

### Option C: OpenAI Fine-tuning (Easiest, paid)
No GPU needed, but data goes to OpenAI

---

### A.1 Install Training Dependencies
```bash
cd ~/Source/Number2
source venv/bin/activate

pip install unsloth
pip install -e ".[training]"
```

### A.2 Verify GPU
```bash
python -c "import torch; print(f'CUDA available: {torch.cuda.is_available()}'); print(f'GPU: {torch.cuda.get_device_name(0) if torch.cuda.is_available() else None}')"
```

### A.3 Run Training
```bash
python scripts/train_lora.py \
    --dataset data/training/training_data_ollama.json \
    --base-model mistral \
    --epochs 3 \
    --output-dir models/style-lora
```

Training will show progress:
```
🚀 Starting training...
   Epochs: 3
   Batch size: 2
   Learning rate: 0.0002

Step 10/300: loss=2.45
Step 20/300: loss=1.89
...
✓ Training complete!
```

### A.4 Import to Ollama
```bash
ollama create my-style -f models/style-lora/Modelfile
ollama list | grep my-style
```

### A.5 Test Fine-tuned Model
```bash
python scripts/test_generation.py \
    --model ollama/my-style \
    --format tweet \
    --topic "Testing my fine-tuned model"
```

---

### B.1 Upload Training Data
Upload `data/training/training_data_ollama.json` to Colab or Google Drive

### B.2 Use Colab Notebook
Create notebook with:
```python
!pip install unsloth

# Then paste contents of scripts/train_lora.py
# Or create a simplified version
```

### B.3 Download Trained Model
After training, download the GGUF file and import to Ollama locally.

---

### C.1 Prepare OpenAI Format
```bash
# Should already exist from Phase 3.1
ls data/training/training_data.jsonl
```

### C.2 Upload and Train
```python
from openai import OpenAI
client = OpenAI()

# Upload file
file = client.files.create(
    file=open("data/training/training_data.jsonl", "rb"),
    purpose="fine-tune"
)

# Start fine-tuning
job = client.fine_tuning.jobs.create(
    training_file=file.id,
    model="gpt-4o-mini-2024-07-18"
)

print(f"Job ID: {job.id}")
```

### C.3 Monitor and Use
```python
# Check status
status = client.fine_tuning.jobs.retrieve(job.id)
print(status.status)

# When complete, model ID is in:
print(status.fine_tuned_model)  # ft:gpt-4o-mini:...:my-style:abc123
```

### C.4 Update Config
Add your fine-tuned model to `configs/models.yaml`

---

## Acceptance Criteria
- [ ] Training completes without errors
- [ ] Model saved/accessible
- [ ] Test generation produces output
- [ ] Quality is noticeably better than RAG-only

## Expected Training Time
| Method | Time | Cost |
|--------|------|------|
| Local (RTX 3080) | 30-60 min | Electricity |
| Colab Free | 1-2 hours | Free |
| OpenAI | 30-60 min | ~$5-20 |

## Troubleshooting

**Out of memory (local)**:
```bash
# Reduce batch size
python scripts/train_lora.py --batch-size 1

# Or use smaller LoRA rank
python scripts/train_lora.py --lora-rank 8
```

**Training loss not decreasing**:
- Check data quality
- Try lower learning rate: `--learning-rate 1e-4`

## Notes
- First training run is experimental - expect to iterate
- Save the model even if quality isn't perfect
- You can retrain with more/better data later
